### ISSUE TITLE
`use` and `label` lead Next to think no response was sent

### ISSUE THREAD

**Issue by Tbhesswebber** [htunnicliff/next-api-middleware#10]
**Describe the bug**

Any _async_ endpoint that I pass middleware to using `use` or `label` ultimately responds to the front-end as expected, but it leads Next to log the below string in development. 

```
API resolved without sending a response for /api/<rest of uri>, this may result in stalled requests.
```

**To Reproduce**
1. Create a function `withMiddleware` as shown below;
2. Create a basic asynchronous handler.
3. Run the application
4. Call the endpoint
5. Observe the logs in the terminal

```
const withMiddleware = label({
  development: [morgan("dev") as ExpressMiddleware],
});

export default withMiddleware(process.env.NODE_ENV)(
  async (req: NextApiRequest, res: NextApiResponse) => {
    await wait(0);
    res.status(200).send([{ value: BookmarkStatus.Saved }]);
  }
);

// helper function
function wait(ms: number = 0) {
  return new Promise(res => setTimeout(res, ms));
}
```

**Expected behavior**

I expect everything to work the way it does, except that Next should process that the handler is working.

You can see the expected behavior by removing the call to `wait` from the sample code.

**Additional context**

I haven't tried any other methods, so I am explicitly calling out the two methods that I have tried

**Comment by htunnicliff** [comment]
Hi @Tbhesswebber – I just attempted to reproduce this example. I think there is a typo in the `wait` function that could be causing this to fail. 

This helper function doesn't actually call the promise resolve argument.
```ts
function wait(ms: number = 0) {
  return new Promise(res => setTimeout(res, ms));
}
```

If you modify this function locally to invoke the resolve argument, the API handler and middleware execute correctly.
```ts
function wait(ms: number = 0) {
  return new Promise(res => {
    setTimeout(() => { 
      res();
    }, ms)
  });
}
```

Do you still see an issue after updating the `wait` function?

**Comment by Tbhesswebber** [comment]
Here's a sandbox that still has the issue after the change to the `wait` function (if that change had fixed things, we'd have found a bug in JavaScript, which would have been cool!).

https://codesandbox.io/s/flamboyant-wind-8ed1e?file=/pages/api/hello.js

**Comment by jorroll** [comment]
Serendipitously, I happened to randomly see this issue right after browsing the Nextjs API documentation. I noticed that the Nextjs API docs for a [`Custom config`](https://nextjs.org/docs/api-routes/api-middlewares#custom-config) happens to include this snippet:

> `externalResolver` is an explicit flag that tells the server that this route is being handled by an external resolver like express or connect. Enabling this option disables warnings for unresolved requests.

It seems possible that this issue is related to that `externalResolver` config option and, if so, the warning should be ignored in the context of using this middleware. How Nextjs can detect that middleware is being used is not clear. It's also not obvious to me how wrapping an API handler in middleware would produce any detectable difference from the framework's perspective.

#### Side note:

```ts
function wait(ms: number = 0) {
  return new Promise(res => setTimeout(res, ms));
}
```

>  I think there is a typo in the wait function that could be causing this to fail. This helper function doesn't actually call the promise resolve argument.

Incorrect. [See the definition of `setTimeout`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout).  `setTimeout(res, ms)` will call the `res` function after a `ms` delay. `res` is just a function in the same way that `() => res()` is just a function.

**Comment by Tbhesswebber** [comment]
> How Nextjs can detect that middleware is being used is not clear. It's also not obvious to me how wrapping an API handler in middleware would produce any detectable difference from the framework's perspective.

To me, this would imply that Next is expecting a promise to be returned (even if that promise resolves to undefined via an implicit return).

I did know about the external resolver property, but part of the reason I went the route of using this package rather than leveraging an express server was because I didn't want to move to a fully managed API and lose any potential optimizations that Next might offer in the future.

**Comment by jorroll** [comment]
> To me, this would imply that Next is expecting a promise to be returned (even if that promise resolves to undefined via an implicit return).

I also thought this might be the case, but I believe the example `hello.js` API route that is created when you initialize a new Nextjs app does not return a promise (and does not log a warning). Additionally, if you change the `hello.js` function into an async function (i.e. force it to return a promise), Nextjs does not log any warnings to the console.

**Comment by Tbhesswebber** [comment]
_This is all based on assumption, so I may be wrong (and will gladly admit it if someone who knows can tell me)._

When a handler isn't async, Next knows if it threw an error because of the single threaded nature of JavaScript. When you make it async, it returns a promise by default (the equivalent of `Promise.resolve(undefined)`) and Next is able to wait for that promise to resolve or reject and handle it appropriately. It's when you get into more complex situations (like piping requests and responses through an array of middleware), things get tricky.

**Comment by htunnicliff** [comment]
> Next is able to wait for that promise to resolve or reject and handle it appropriately. 

I think the problem is [here](https://github.com/htunnicliff/next-api-middleware/blob/canary/lib/index.ts#L88-L109):
```ts
export function makeMiddlewareExecutor(middlewareFns: Middleware[]) {
  return function curryApiHandler(apiRouteFn: NextApiHandler): NextApiHandler {
    return async function finalRouteHandler(req, res) {
      const execute = ([
        currentFn,
        ...remaining
      ]: Middleware[]): void | Promise<void> => {
        return currentFn(req, res, async (err?: any) => { // <-------- PROBLEMATIC LINE
          if (err) {
            throw err;
          }

          return remaining.length === 0
            ? apiRouteFn(req, res)
            : execute(remaining);
        });
      };

      return execute(middlewareFns);
    };
  };
}
```

If the first middleware function invoked is not async, that prevents Next.js from knowing that the request was resolved correctly. 

I think the solution here will be a modification of the source code that ensures that `currentFn` acts asynchronously.

**Comment by htunnicliff** [comment]
I took a crack at refactoring how middleware gets executed, and I believe this may handle callback-style middleware and async middleware correctly.

Would you mind cloning and seeing if it resolves the error you've encountered? @Tbhesswebber @jorroll 

PR: https://github.com/htunnicliff/next-api-middleware/pull/12

**Comment by jorroll** [comment]
@htunnicliff sorry for the miscommunication. I've never encountered this error with `next-api-middleware`. I simply noticed this issue and was reminded of something I read in the [nextjs API docs](https://github.com/htunnicliff/next-api-middleware/issues/10#issuecomment-860177159) so figured I'd share the information.

**Comment by htunnicliff** [comment]
@Tbhesswebber this bug should be resolved in [v1.0.0](https://github.com/htunnicliff/next-api-middleware/releases/tag/1.0.0).
